src/libraries/System.Private.CoreLib/src/System/DateTime.cs

```diff
@@ -254,52 +240,41 @@ public DateTime(int year, int month, int day, int hour, int minute, int second,
         //
         public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond)
         {
-            if (millisecond < 0 || millisecond >= MillisPerSecond)
+            if ((uint)millisecond >= MillisPerSecond) ThrowMillisecondOutOfRange();
+
+            if (second != 60 || !s_systemSupportsLeapSeconds)
             {
-                throw new ArgumentOutOfRangeException(nameof(millisecond), SR.Format(SR.ArgumentOutOfRange_Range, 0, MillisPerSecond - 1));
+                ulong ticks = DateToTicks(year, month, day) + TimeToTicks(hour, minute, second);
+                ticks += (uint)millisecond * (uint)TicksPerMillisecond;
+                Debug.Assert(ticks <= MaxTicks, "Input parameters validated already");
+                _dateData = ticks;
             }
-
-            if (second == 60 && s_systemSupportsLeapSeconds && IsValidTimeWithLeapSeconds(year, month, day, hour, minute, second, DateTimeKind.Unspecified))
+            else
             {
-                // if we have leap second (second = 60) then we'll need to check if it is valid time.
-                // if it is valid, then we adjust the second to 59 so DateTime will consider this second is last second
-                // in the specified minute.
-                // if it is not valid time, we'll eventually throw.
-                second = 59;
+                // if we have a leap second, then we adjust it to 59 so that DateTime will consider it the last in the specified minute.
+                this = new(year, month, day, hour, minute, 59, 999);
```

Comment:
This part need to be fixed. pass the input millisecond instead of 999

Reply:
Converting 60.nnn into 59.nnn sounds like the worst possible option (others being 59.999 or 59.9999999). From the original code structure it looked like an oversight not an intentional decision?

Reply:
> Converting 60.nnn into 59.nnn sounds like the worst possible option (others being 59.999 or 59.9999999). From the original code structure it looked like an oversight not an intentional decision?

You never know what is the intentional of the constructor caller. At least we are sure this is not coming from `UtcNow` as we handle that there. We change the second to 59 because we had no other options and it is easy to explain. Changing the milliseconds can be confusing especially if this is not desired by the caller.

Reply:
Thanks for addressing the feedback!